File: app/src/main/java/me/destro/android/gitfav/features/listing/adapters/StarredRepositoriesAdapter.java
package me.destro.android.gitfav.features.listing.adapters;

import androidx.databinding.DataBindingUtil;
import androidx.paging.PagedListAdapter;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import me.destro.android.gitfav.R;
import me.destro.android.gitfav.databinding.ItemGithubRepositoryBinding;
import me.destro.android.libraries.github.model.StarredRepository;
import me.destro.android.gitfav.features.listing.paging.StarredRepositoryDiffCallback;

public class StarredRepositoriesAdapter extends PagedListAdapter<StarredRepository, StarredRepositoriesAdapter.StarredRepositoryHolder> {

    public StarredRepositoriesAdapter() {
        super(new StarredRepositoryDiffCallback());
    }

    @NonNull
    @Override
    public StarredRepositoryHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View root = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_github_repository, viewGroup, false);
        return new StarredRepositoryHolder(root);
    }

    @Override
    public void onBindViewHolder(@NonNull StarredRepositoryHolder viewHolder, int i) {
        StarredRepository item = getItem(i);
        if (item != null)
            viewHolder.bind(item);
    }

    class StarredRepositoryHolder extends RecyclerView.ViewHolder {

        private ItemGithubRepositoryBinding binding;

        private TextView fullProjectName;
        private TextView description;
        private TextView stargazersCount;
        private ChipGroup topicsChipGroup;

        public StarredRepositoryHolder(@NonNull View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
            fullProjectName = binding.fullProjectName;
            description = binding.description;
            stargazersCount = binding.stargazersCount;
            topicsChipGroup = binding.topics;
        }

        void bind(@NonNull StarredRepository starredRepository) {
            fullProjectName.setText(starredRepository.fullName);
            description.setText(starredRepository.description);
            stargazersCount.setText(String.valueOf(starredRepository.stargazersCount));

            if (starredRepository.topics.length != 0) {
                buildTopicsChips(starredRepository.topics);
                topicsChipGroup.setVisibility(View.VISIBLE);
            } else {
                topicsChipGroup.setVisibility(View.GONE);
            }
        }

        private void buildTopicsChips(String[] topics) {
            // Create new child as needed.
            int childCountBefore = topicsChipGroup.getChildCount();
            for (int i = 0; i < topics.length - childCountBefore; i++) {
                topicsChipGroup.addView(new Chip(topicsChipGroup.getContext()));
            }


            for (int i = 0; i < topics.length; i++) {
                String topic = topics[i];
                ((Chip)topicsChipGroup.getChildAt(i)).setText(topic);
            }

            // Hide those not-used chips.
            for (int i = topics.length; i < topicsChipGroup.getChildCount(); i++) {
                topicsChipGroup.getChildAt(i).setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/me/destro/android/gitfav/features/listing/adapters/StarredRepositoriesAdapter.kt
package me.destro.android.gitfav.features.listing.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.databinding.DataBindingUtil
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import me.destro.android.gitfav.R
import me.destro.android.gitfav.databinding.ItemGithubRepositoryBinding
import me.destro.android.gitfav.domain.model.Repository
import me.destro.android.gitfav.features.listing.paging.StarredRepositoryDiffCallback
import java.io.IOException

class StarredRepositoriesAdapter : PagedListAdapter<Repository, StarredRepositoriesAdapter.StarredRepositoryHolder>(StarredRepositoryDiffCallback()) {

    private var onStarredRepositoryClickListener: ((Repository) -> Unit)? = null

    fun setOnStarredRepositoryClickListener(onRepositoryClickListener: (Repository) -> Unit) {
        this.onStarredRepositoryClickListener = onRepositoryClickListener
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): StarredRepositoryHolder {
        val root = LayoutInflater.from(viewGroup.context).inflate(R.layout.item_github_repository, viewGroup, false)
        return StarredRepositoryHolder(root)
    }

    override fun onBindViewHolder(viewHolder: StarredRepositoryHolder, i: Int) {
        val item = getItem(i)
        if (item != null)
            viewHolder.bind(item)
    }

    inner class StarredRepositoryHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val binding: ItemGithubRepositoryBinding = DataBindingUtil.bind(itemView)!!

        private val fullProjectName: TextView
        private val description: TextView
        private val stargazersCount: TextView
        private val topicsChipGroup: ChipGroup
        private val languagesChipGroup: ChipGroup
        private val userTagsChipGroup: ChipGroup

        init {
            fullProjectName = binding.fullProjectName
            description = binding.description
            stargazersCount = binding.stargazersCount
            topicsChipGroup = binding.topics
            languagesChipGroup = binding.languages
            userTagsChipGroup = binding.userTags
        }

        fun bind(starredRepository: Repository) {
            fullProjectName.text = starredRepository.name

            if (starredRepository.description != null) {
                description.text = starredRepository.description
            }else {
                description.visibility = View.GONE
            }

            stargazersCount.text = starredRepository.starsCount.toString()

            if (starredRepository.topics.isNotEmpty()) {
                buildTopicsChips(starredRepository.topics)
                topicsChipGroup.visibility = View.VISIBLE
            }else {
                topicsChipGroup.visibility = View.GONE
            }

            if (starredRepository.languages.isNotEmpty()) {
                languagesChipGroup.visibility = View.VISIBLE
            }else {
                languagesChipGroup.visibility = View.GONE
            }

            if (starredRepository.userTags.isNotEmpty()) {
                userTagsChipGroup.visibility = View.VISIBLE
            }else {
                userTagsChipGroup.visibility = View.GONE
            }

            if (onStarredRepositoryClickListener != null) {
                binding.root.setOnClickListener { _ -> onStarredRepositoryClickListener?.invoke(starredRepository) }
            }
        }

        private fun buildTopicsChips(topics: Array<String>) {
            // Create new child as needed.
            val childCountBefore = topicsChipGroup.childCount
            for (i in 0 until topics.size - childCountBefore) {
                topicsChipGroup.addView(Chip(topicsChipGroup.context))
            }

            for (i in topics.indices) {
                val topic = topics[i]
                (topicsChipGroup.getChildAt(i) as Chip).text = topic
            }

            // Hide those not-used chips.
            for (i in topics.size until topicsChipGroup.childCount) {
                topicsChipGroup.getChildAt(i).visibility = View.GONE
            }
        }
    }
}